doc/poddoc/man/btexec.8
.\" Automatically generated by Pod::Man 2.25 (Pod::Simple 3.16)
.\"
.\" Standard preamble:
.\" ========================================================================
.de Sp \" Vertical space (when we can't use .PP)
.if t .sp .5v
.if n .sp
..
.de Vb \" Begin verbatim text
.ft CW
.nf
.ne \\$1
..
.de Ve \" End verbatim text
.ft R
.fi
..
.\" Set up some character translations and predefined strings.  \*(-- will
.\" give an unbreakable dash, \*(PI will give pi, \*(L" will give a left
.\" double quote, and \*(R" will give a right double quote.  \*(C+ will
.\" give a nicer C++.  Capital omega is used to do unbreakable dashes and
.\" therefore won't be available.  \*(C` and \*(C' expand to `' in nroff,
.\" nothing in troff, for use with C<>.
.tr \(*W-
.ds C+ C\v'-.1v'\h'-1p'\s-2+\h'-1p'+\s0\v'.1v'\h'-1p'
.ie n \{\
.    ds -- \(*W-
.    ds PI pi
.    if (\n(.H=4u)&(1m=24u) .ds -- \(*W\h'-12u'\(*W\h'-12u'-\" diablo 10 pitch
.    if (\n(.H=4u)&(1m=20u) .ds -- \(*W\h'-12u'\(*W\h'-8u'-\"  diablo 12 pitch
.    ds L" ""
.    ds R" ""
.    ds C` ""
.    ds C' ""
'br\}
.el\{\
.    ds -- \|\(em\|
.    ds PI \(*p
.    ds L" ``
.    ds R" ''
'br\}
.\"
.\" Escape single quotes in literal strings from groff's Unicode transform.
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\"
.\" If the F register is turned on, we'll generate index entries on stderr for
.\" titles (.TH), headers (.SH), subsections (.SS), items (.Ip), and index
.\" entries marked with X<> in POD.  Of course, you'll have to process the
.\" output yourself in some meaningful fashion.
.ie \nF \{\
.    de IX
.    tm Index:\\$1\t\\n%\t"\\$2"
..
.    nr % 0
.    rr F
.\}
.el \{\
.    de IX
..
.\}
.\"
.\" Accent mark definitions (@(#)ms.acc 1.5 88/02/08 SMI; from UCB 4.2).
.\" Fear.  Run.  Save yourself.  No user-serviceable parts.
.    \" fudge factors for nroff and troff
.if n \{\
.    ds #H 0
.    ds #V .8m
.    ds #F .3m
.    ds #[ \f1
.    ds #] \fP
.\}
.if t \{\
.    ds #H ((1u-(\\\\n(.fu%2u))*.13m)
.    ds #V .6m
.    ds #F 0
.    ds #[ \&
.    ds #] \&
.\}
.    \" simple accents for nroff and troff
.if n \{\
.    ds ' \&
.    ds ` \&
.    ds ^ \&
.    ds , \&
.    ds ~ ~
.    ds /
.\}
.if t \{\
.    ds ' \\k:\h'-(\\n(.wu*8/10-\*(#H)'\'\h"|\\n:u"
.    ds ` \\k:\h'-(\\n(.wu*8/10-\*(#H)'\`\h'|\\n:u'
.    ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'^\h'|\\n:u'
.    ds , \\k:\h'-(\\n(.wu*8/10)',\h'|\\n:u'
.    ds ~ \\k:\h'-(\\n(.wu-\*(#H-.1m)'~\h'|\\n:u'
.    ds / \\k:\h'-(\\n(.wu*8/10-\*(#H)'\z\(sl\h'|\\n:u'
.\}
.    \" troff and (daisy-wheel) nroff accents
.ds : \\k:\h'-(\\n(.wu*8/10-\*(#H+.1m+\*(#F)'\v'-\*(#V'\z.\h'.2m+\*(#F'.\h'|\\n:u'\v'\*(#V'
.ds 8 \h'\*(#H'\(*b\h'-\*(#H'
.ds o \\k:\h'-(\\n(.wu+\w'\(de'u-\*(#H)/2u'\v'-.3n'\*(#[\z\(de\v'.3n'\h'|\\n:u'\*(#]
.ds d- \h'\*(#H'\(pd\h'-\w'~'u'\v'-.25m'\f2\(hy\fP\v'.25m'\h'-\*(#H'
.ds D- D\\k:\h'-\w'D'u'\v'-.11m'\z\(hy\v'.11m'\h'|\\n:u'
.ds th \*(#[\v'.3m'\s+1I\s-1\v'-.3m'\h'-(\w'I'u*2/3)'\s-1o\s+1\*(#]
.ds Th \*(#[\s+2I\s-2\h'-\w'I'u*3/5'\v'-.3m'o\v'.3m'\*(#]
.ds ae a\h'-(\w'a'u*4/10)'e
.ds Ae A\h'-(\w'A'u*4/10)'E
.    \" corrections for vroff
.if v .ds ~ \\k:\h'-(\\n(.wu*9/10-\*(#H)'\s-2\u~\d\s+2\h'|\\n:u'
.if v .ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'\v'-.4m'^\v'.4m'\h'|\\n:u'
.    \" for low resolution devices (crt and lpr)
.if \n(.H>23 .if \n(.V>19 \
\{\
.    ds : e
.    ds 8 ss
.    ds o a
.    ds d- d\h'-1'\(ga
.    ds D- D\h'-1'\(hy
.    ds th \o'bp'
.    ds Th \o'LP'
.    ds ae ae
.    ds Ae AE
.\}
.rm #[ #] #H #V #F C
.\" ========================================================================
.\"
.IX Title "BTEXEC 8"
.TH BTEXEC 8 "2009-05-18" "GNUbatch Release 1" "GNUbatch Batch Scheduler"
.\" For nroff, turn off justification.  Always turn off hyphenation; it makes
.\" way too many mistakes in technical documents.
.if n .ad l
.nh
.SH "NAME"
btexec \- run commands for macros
.SH "SYNOPSIS"
.IX Header "SYNOPSIS"
\&\fB/usr/spool/progs/btexec\fR
\&\fIoptions\fR
.SH "DESCRIPTION"
.IX Header "DESCRIPTION"
\&\fBBtexec\fR runs commands for macros under the identity of the invoking
user. This is required because \fIgbch\-q\fR\|(1), \fIgbch\-xq\fR\|(1) and \fIgbch\-xmq\fR\|(1)
are set-user programs (to other than \f(CW\*(C`root\*(C'\fR) and there is an inherent
security breach in many versions of Unix in that such programs cannot
divest themselves of traces of the set-user user id.
.PP
This program is only intended for internal use and is not further
documented.
.SH "COPYRIGHT"
.IX Header "COPYRIGHT"
Copyright (c) 2009 Free Software Foundation, Inc.
This is free software. You may redistribute copies of it under the
terms of the \s-1GNU\s0 General Public License
<http://www.gnu.org/licenses/gpl.html>.
There is \s-1NO\s0 \s-1WARRANTY\s0, to the extent permitted by law.
.SH "AUTHOR"
.IX Header "AUTHOR"
John M Collins, Xi Software Ltd.


debian/changelog
gnubatch (1.5) unstable; urgency=low

  * Release 1.5
    Revise network handling, increase user/host lengths in API.
    Mostly rewrite xbnetserv and half of sh_network.c
    Fix bug with expansion of names in redirections.

 -- John M Collins (Xi Software Ltd) <jmc@xisl.com>  Tue, 26 Jun 2012 12:45:06 +0100

gnubatch (1.1) unstable; urgency=low

  * Initial release (Closes: #572633)

 -- John M Collins (Xi Software Ltd) <jmc@xisl.com>  Fri, 05 Mar 2010 11:22:06 +0000



src/btexec.c
/* btexec.c -- main module for btexec

   Copyright 2009 Free Software Foundation, Inc.

   This program is free software: you can redistribute it and/or modify
   it under the terms of the GNU General Public License as published by
   the Free Software Foundation, either version 3 of the License, or
   (at your option) any later version.

   This program is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
   GNU General Public License for more details.

   You should have received a copy of the GNU General Public License
   along with this program.  If not, see <http://www.gnu.org/licenses/>.  */

#include "config.h"
#include <sys/types.h>
#include <sys/stat.h>
#include "incl_unix.h"
#include <stdio.h>
#include <pwd.h>
#include "defaults.h"
#include "incl_unix.h"
#include "files.h"
#include "ecodes.h"

/* This is a helper program for macros in user interface stuff so
   that macro commands get run under the identity of the user who
   invoked the original UI command.

   It has to be run set-uid "root" so that setuid(x) eliminates
   all traces of any set-user stuff.

   Originally it just did setuid(getuid()) but that doesn't work
   for GTK stuff which won't work set-user. To fool GTK you have
   to do setreuid to batch losing the real uid. So what we do
   in the GTK programs is check that $HOME is owned by the real
   user and we set-user to the owner of the HOME directory here.

   I think this is secure but if it isn't please let me know
   ASAP!!! */


/* Rewrite the first argument of the argument list so that
   things like btjchange -args work in macros. (Otherwise the
   macro command tries to run "spchange -args" as a command name). */

char **rewrite(char **argv)
{
	unsigned  nargs = 1;	/* Including null */
	char	**ap, *sp, *np;
	char	**result, **rp;

	/* If it doesn't have spaces in, nothing to do */

	if  (!strchr(argv[0], ' '))
		return  argv;

	/* Count the number of args */

	for  (ap = argv;  *ap;  ap++)
		nargs++;

	/* Assume increase in number of args is same as number of spaces
	   this may be an overestimate */

	for  (sp = argv[0];  (np = strchr(sp, ' '));  sp = np + 1)
		nargs++;

	result = (char **) malloc(nargs * sizeof(char *));
	if  (!result)
		exit(E_NOMEM);

	rp = result;

	for  (sp = argv[0];  (np = strchr(sp, ' '));  sp = np)  {
		*np = '\0';	/* Assume doesn't matter if we clobber it */
		*rp++ = sp;
		do  np++;
		while  (*np == ' ');
	}

	/* Add trailing stuff */

	if  (*sp)
		*rp++ = sp;

	/* Add rest of args */

	for  (ap = argv + 1;  *ap;  ap++)
		*rp++ = *ap;
	*rp = (char *) 0;

	return  result;
}

MAINFN_TYPE  main(int argc, char **argv)
{
	struct  passwd  *pw = getpwnam(BATCHUNAME);
	uid_t  duid = ROOTID, myuid = getuid();
	char	**newargv;

	versionprint(argv, "$Revision: 1.5 $", 1);

	/* Get batch uid */

	if  (pw)
		duid = pw->pw_uid;
	endpwent();

	/* If the real user id is "batch" this is either because we have invoked the
	   original ui program as batch or because we have invoked a GTK program which
	   switched the real uid to batch. In such cases we fish the uid out of the
	   home directory and use that.

	   Refuse to work if no home directory or it looks strange */

	if  (myuid == duid)  {
		char	*homed = getenv("HOME");
		struct  stat  sbuf;

		if  (!homed  ||  stat(homed, &sbuf) < 0  ||  (sbuf.st_mode & S_IFMT) != S_IFDIR)
			exit(E_SETUP);

		myuid = sbuf.st_uid;
	}

	/* Switch completely to the user id and do the bizniz. */

	setuid(myuid);
	newargv = rewrite(argv);
	execvp(newargv[0], newargv);
	exit(E_BTEXEC2);
}
